onboard/catkin_ws/src/controls/scripts/desired_state.py

```diff
@@ -45,40 +43,77 @@ def __init__(self):
         self.listener = TransformListener()
         self.pid_manager = PIDManager()
 
-        rospy.Subscriber(self.DESIRED_POSE_TOPIC, Pose, self._on_pose_received)
-        rospy.Subscriber(self.DESIRED_TWIST_TOPIC,
-                         Twist, self._on_twist_received)
-        rospy.Subscriber(self.DESIRED_POWER_TOPIC,
-                         Twist, self._on_power_received)
-
-    def _on_pose_received(self, pose):
+        self.pose_server = actionlib.SimpleActionServer(
+            self.DESIRED_POSE_ACTION,
+            ControlsDesiredPoseAction,
+            self._on_pose_received,
+            False)
+        self.twist_server = actionlib.SimpleActionServer(
+            self.DESIRED_TWIST_ACTION,
+            ControlsDesiredTwistAction,
+            self._on_twist_received,
+            False)
+        self.power_server = actionlib.SimpleActionServer(
+            self.DESIRED_POWER_ACTION,
+            ControlsDesiredPowerAction,
+            self._on_power_received,
+            False)
+
+        self.pose_server.start()
+        self.twist_server.start()
+        self.power_server.start()
+
+    def _on_pose_received(self, goal):
         """Handler for receiving desired pose. Transforms global desired pose to local reference frame
         for PID loops.
 
         Args:
             pose: ROS Pose message corresponding to desired pose in global reference frame
         """
-        self.pose = utils.parse_pose(utils.transform_pose(
-            self.listener, 'odom', 'base_link', pose))
+        self.twist_server.preempt_request = True
+        self.power_server.preempt_request = True
+
+        pose = utils.parse_pose(utils.transform_pose(
+            self.listener, 'odom', 'base_link', goal.pose))
+
+        self._publish_control(self.pose_server, self.pid_manager.position_control, pose)
 
-    def _on_twist_received(self, twist):
+    def _on_twist_received(self, goal):
         """Handler for receiving desired twists. Received desired twists are assumed to be defined in the
         local reference frame.
 
         Args:
             power: ROS Twist message corresponding to desired twists in local reference frame
         """
-        self.twist = utils.parse_twist(twist)
+        self.pose_server.preempt_request = True
+        self.power_server.preempt_request = True
 
-    def _on_power_received(self, power):
+        twist = utils.parse_twist(goal.twist)
+
+        if not self._twist_state_safety(twist):
+            self.twist_server.set_aborted()
```

Comment:
Can we log an error message here

Reply:
`_twist_state_safety` already logs for each axis that is out of bounds